File: generate.py
import pickle
import random
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
import matplotlib.animation as animation
from tqdm import tqdm

# "Mobility promotes and jeopardizes biodiversity in rock–paper–scissors games." Reichenbach, Mobilia & Frey (2007)
FFMPEG_PATH = "/usr/local/bin/ffmpeg"


class Grid:
    # Defines the function for reiterating time steps, interaction methods, and image/video processing
    def __init__(self, seed, n, max_steps, one_time_step):
        self.seed = seed
        self.length = n
        self.num_cells = n ** 2
        self.max_steps = max_steps
        self.current_step_count = 0  # Counts the number of time steps
        self.cell_status = [1, 2, 3]  # 1= type A(red), 2= type B(blue), 3= type C(yellow)
        self.periodic_boundary_on = False
        self.ising_on = True  # TRUE/FALSE for biased diffusion
        self.one_time_step = one_time_step

        self.reproduction_param = 1  # A + 0 --> A + A
        self.selection_param = 1  # A + B --> A + 0
        self.exchange_param = 1.e-6 * self.num_cells / 2  # M = 2 * exchange_param / number of sites
        # Newly introduced parameters
        self.ising_param = 1.2
        self.death_param = 0.6
        # Additional parameters set in Szczesny, Mobilia & Rucklidge (2014)
        self.replacement_param = 0  # A + B --> A + A
        self.mutation_param = 0  # A --> B

        # Average probability of an interaction in the next time step dt
        self.dispatcher = [self.exchange]
        self.pairings = {"death": self.death_param,
                         "exchange": self.exchange_param,
                         "mutation": self.mutation_param,
                         "selection": self.selection_param,
                         "replacement": self.replacement_param,
                         "reproduction": self.reproduction_param}

        diff_prob = [self.pairings[d.__name__] for d in self.dispatcher]
        total_prob = sum(diff_prob)
        self.diff_prob = np.divide(diff_prob, total_prob)

        self.current_cell = ()
        self.current_neighbours = []
        self.img_frames = []  # Stores the grid array at every time step
        self.first_frame = None  # Initialises the first frame for animation later

    def call(self):
        # Reiterates each time step until the step limit has been reached
        for i in tqdm(range(self.max_steps), desc="Progress bar"):
            self.one_time_step()
            self.current_step_count = i

        print("Simulation complete, now loading animation...")

    def call_neighbours(self, cell_coord):
        x, y = cell_coord[0], cell_coord[1]
        if self.periodic_boundary_on:
            self.current_neighbours = [(x, (y + 1) % self.length),
                                       (x, (y - 1) % self.length),
                                       ((x + 1) % self.length, y),
                                       ((x - 1) % self.length, y)]
        else:
            neighbours = [(x, (y + 1)), (x, (y - 1)), ((x + 1), y), ((x - 1), y)]
            self.current_neighbours = [(x, y) for (x, y) in neighbours
                                       if 0 <= x < self.length and 0 <= y < self.length]
        return self.current_neighbours

    def calc_neighbours(self, grid, neighbours):
        bias = len([i for i in neighbours if grid[i] > 0])
        return bias

    def exchange(self, grid, cell, neighbour):
        # Swaps positions with a neighbouring cell
        calc_original = self.calc_neighbours(grid, self.current_neighbours)
        grid[cell], grid[neighbour] = grid[neighbour], grid[cell]
        if self.ising_on and grid[cell] == 0:
            calc_new = self.calc_neighbours(grid, self.call_neighbours(neighbour))
            num_J = calc_original - calc_new
            prob = min(1, np.exp(- num_J * self.ising_param))
            if prob < np.random.rand():
                grid[cell], grid[neighbour] = grid[neighbour], grid[cell]

    def selection(self, grid, cell, neighbour):
        # Destroys a neighbouring cell via the "rock-paper-scissors" mechanic
        # Type A destroys B, type B destroys C, type C destroys A
        cell_index = (self.cell_status.index(grid[cell]) + 1) % len(self.cell_status)
        if grid[neighbour] == self.cell_status[cell_index]:
            grid[neighbour] = 0

    def reproduction(self, grid, cell, neighbour):
        # Generates a new cell of the same type in the neighbouring position
        if grid[neighbour] == 0:
            grid[neighbour] = grid[cell]

    def death(self, grid, cell, _):
        # Cell leaves a vacant position after death
        grid[cell] = 0

    def replacement(self, grid, cell, neighbour):
        # Replaces a neighbouring cell via the "rock-paper-scissors" mechanic
        # Based off dominance-replacement in Szczesny, Mobilia & Rucklidge (2014)
        cell_index = (self.cell_status.index(grid[cell]) + 1) % len(self.cell_status)
        if grid[neighbour] == self.cell_status[cell_index]:
            grid[neighbour] = grid[cell]

    def mutation(self, grid, cell, _):
        # Mutates one cell type into the other
        cell_index = (self.cell_status.index(grid[cell]) + 1) % len(self.cell_status)
        grid[cell] = self.cell_status[cell_index]

    def save_image(self):
        # Saves each successive iteration of the grid as a .png
        for ind, grid in enumerate(self.img_frames):
            cmap = ListedColormap(["black", "red", "blue", "yellow"])
            plt.imshow(grid, cmap=cmap, interpolation='nearest')
            plt.savefig(f"frame_{ind}.png")

    def save_video(self, movie_path="./output/movie.mp4", ffmpeg_path=None):
        # Generates a .mp4 from the time evolution of the grid
        fig, ax = plt.subplots()
        cmap = ListedColormap(["black", "red", "blue", "yellow"])
        self.first_frame = ax.imshow(self.img_frames[0], cmap=cmap, interpolation='nearest', animated=True)

        if ffmpeg_path:
            plt.rcParams["animation.ffmpeg_path"] = ffmpeg_path

        anim = animation.FuncAnimation(fig, self.animate, frames=len(self.img_frames))
        anim.save(movie_path, writer=animation.FFMpegWriter(fps=60))
        plt.close()

    def save_state(self, state_path="./output/state.pickle"):
        state = {}
        prop_full, prop_dead = [], []

        for grid in self.img_frames:
            ar_unique, i = np.unique(grid, return_counts=True)
            prop_full.append(i[1:] / self.num_cells)
            prop_dead.append(i[0] / self.num_cells)

        state["var"] = {"seed": self.seed,
                        "system_size": self.num_cells,
                        "time_steps": self.max_steps}

        state["param"] = {"exchange": self.exchange_param,
                          "death": self.death_param,
                          "ising": self.ising_param,
                          "mutation": self.mutation_param,
                          "replacement": self.replacement_param,
                          "reproduction": self.reproduction_param,
                          "selection": self.selection_param}

        state["data"] = {"prop_full": prop_full,
                         "prop_dead": prop_dead,
                         "time_evol": self.img_frames}

        with open(state_path, "wb") as file:
            pickle.dump(state, file)

    def animate(self, frame):
        next_frame = self.img_frames[frame]
        self.first_frame.set_array(next_frame)
        return next_frame


class PlainGrid(Grid):
    # Defines the initial grid set-up and each time step
    def __init__(self, n, max_steps, seed=12345):
        Grid.__init__(self, seed, n, max_steps, self.one_time_step)
        self.grid = self.initial_grid()

    def initial_grid(self):
        np.random.seed(self.seed)  # Initialises the randomly populated grid with a fixed seed
        grid = np.random.choice(4, size=(self.length, self.length), p=[0.76, 0.08, 0.08, 0.08])
        # grid = np.pad(grid, pad_width=1, mode='constant', constant_values=0)  # Adds an empty border
        return grid

    def one_time_step(self):
        reaction_count = 0
        while reaction_count < self.num_cells:
            cell_coord = tuple(np.random.choice(self.length, size=2))  # Randomly chooses cell coordinates
            # Only valid if value at the chosen coordinates != 0
            if self.grid[cell_coord]:
                neighbour_coord = random.choice(self.call_neighbours(cell_coord))  # Randomly chooses one neighbour
                reaction = np.random.choice(len(self.dispatcher), p=self.diff_prob)  # Randomly chooses one event
                self.dispatcher[reaction](self.grid, cell_coord, neighbour_coord)
                reaction_count += 1

        self.img_frames.append(np.copy(self.grid))

File: generate.jl
using Pkg
Pkg.add("StatsBase")
Pkg.add("DataStructures")
Pkg.add("JLD2")
Pkg.add("ProgressBars")
using StatsBase, DataStructures, JLD2, ProgressBars
using Random, Plots, FixedPointNumbers

# One-liners
x ± y = (x-y, x+y)
unzip(a) = map(x->getfield.(a, x), fieldnames(eltype(a)))

# Definition of interaction events
function call_neighbours(x, y, n)
    neighbours = [(x, mod1(y + 1, n)),
                  (x, mod1(y - 1, n)),
                  (mod1(x + 1, n), y),
                  (mod1(x - 1, n), y)]
    if ! periodic_boundary_on
        neighbours = [(w, z) for (w, z) in neighbours if w in x ± 1 || z in y ± 1]
    else
        return neighbours
    end
end

function calc_neighbours(grid::Array, neighbours::Vector)
    num_full = sum([1 for (x, y) in neighbours if grid[x, y] > 0])
end

function death(grid::Array, x::Int, y::Int, args...)
    grid[x, y] = 0
end

function exchange(grid::Array, x::Int, y::Int, new_x::Int, new_y::Int, ising_on::Bool)
    if ! ising_on || grid[new_x, new_y] > 0
        grid[x, y], grid[new_x, new_y] = grid[new_x, new_y], grid[x, y]
    else
        n, n = size(grid)
        calc_original = calc_neighbours(grid, call_neighbours(x, y, n))
        calc_new = calc_neighbours(grid, call_neighbours(new_x, new_y, n)) - 1
        if rand() < ising(calc_original, calc_new)
            grid[x, y], grid[new_x, new_y] = grid[new_x, new_y], grid[x, y]
        end
    end
end

function ising(calc_original::Int, calc_new::Int)
    Δ = calc_original - calc_new
    prob = min(1, exp(- 2 * Δ * pairings[ising]))
end

function reproduction(grid::Array, x::Int, y::Int, new_x::Int, new_y::Int)
    if grid[new_x, new_y] == 0
        grid[new_x, new_y] = grid[x, y]
    end
end

function selection(grid::Array, x::Int, y::Int, new_x::Int, new_y::Int)
    if grid[new_x, new_y] == grid[x, y] % 3 + 1
        grid[new_x, new_y] = 0
    end
end

# Bulk flow of Monte Carlo simulations
pairings = Dict{Function, Real}(death => 0,
                                exchange => 1,
                                ising => 0.7,
                                reproduction => 0,
                                selection => 0)
periodic_boundary_on = false

function to_colour(i::Int)
    if i == 0
        return colorant"black"
    else
        return colorant"red"
    end
end
    # elseif i == 1
        # return colorant"red"
    # elseif i == 2
        # return colorant"blue"
    # elseif i == 3
        # return colorant"yellow


function initial_grid(n::Int, default_seed::Int=123456)
    Random.seed!(default_seed)
    sample(0:3, Weights([6, 1, 1, 1]), (n, n))
end

function call(n::Int, max_steps::Int, filename::String="")
    tot_cells = n * n
    grid = initial_grid(n)
    img_frames = [to_colour.(grid)]
    new_pairings = [(a, b) for (a, b) in pairings if b > 0 && a != ising]
    active_events, probabilities = unzip(new_pairings)
    probabilities = probabilities / sum(probabilities)
    ising_on = pairings[ising] > 0

    for time_step ∈ ProgressBar(1:max_steps)
        for cell_step ∈ 1:tot_cells
            x, y = rand(1:n, 2)
            if grid[x, y] > 0
                new_x, new_y = rand(call_neighbours(x, y, n))
                event = sample(active_events, Weights(probabilities))
                ! (event == exchange) ? event(grid, x, y, new_x, new_y) :
                                        event(grid, x, y, new_x, new_y, ising_on)
            end
        ! isempty(filename) && push!(img_frames, to_colour.(grid))
        end
    end

    # Saving the raw data for visualisation later
    println("Simulation complete, compiling data...")
    if ! isempty(filename)
        img_frames = cat(dims=3, img_frames...)
        save(filename * ".gif", img_frames, fps=5)
        jldsave(filename * ".jld2"; grid_length=n, time_steps=max_steps, parameters=pairings, time_evol=img_frames)
        println("Animation saved")
    end
    return to_colour.(grid)
end

# Extract data from .jld2 files
function summarise(filename::String)
    file_data = load(filename)
    n, max_steps = file_data["grid_length"], file_data["time_steps"]
    parameters = file_data["parameters"]
    all_frames = file_data["time_evol"]
    tot_cells = n * n

    println("########################################")
    println("$n x $n grid after $max_steps time steps.")
    println("########################################")
    println("Parameters (", length(parameters), " entries):")
    for (k, v) in parameters
        println("$k => $v")
    end
    println("########################################")
    last_frame = all_frames[:, :, end]
    empty, c, b, a = values(counter(last_frame))
    empty_p, c_p, b_p, a_p = round.([empty, c, b, a] * 100 / tot_cells, sigdigits=3)

    println("Final numbers of different particles:")
    println("Species A => $a ($a_p%)")
    println("Species B => $b ($b_p%)")
    println("Species C => $c ($c_p%)")
    println("Empty => $empty ($empty_p%)")
    println("########################################")
end

